content/profiling_interface.tex

```diff
@@ -0,0 +1,196 @@
+\VAR{BUILD\_PSHMEM\_INTERFACES} macro symbol.
+
+\subsection{Limitations}
+\label{sec:pshmem_limitations}
+
+\subsubsection{Multiple Counting}
+\label{sec:pshmem_multiple_count}
+Since some functions in \openshmem library may itself be implemented 
+using more basic \openshmem functions, it is possible for these 
+profiling functions to be called from within an \openshmem function 
+that was originally called from a profiling routine. For example, 
+\openshmem collective operations can be implemented using basic 
+point-to-point operations. Thus, profiling such a collective 
+operation may lead to counting a profiling function for 
+point-to-point operation more than once after being called from the 
+collective function. It is the developer's responsibility to protect 
+the profiling application not to count a function more than once if 
+that effect is not intended. For a single-threaded profiler, this 
+can be achieved through a static variable counting the number of 
+times a function has been profiled. In a multi-threaded environment, 
+additional synchronizations are needed to manage updates to this 
+counter and thus, becomes more complex to accurately profile the 
+\openshmem functions.
+
+\subsubsection{Separate Build and Link}
+\label{sec:pshmem_separate_build_link}
+To build the profiling tool with both the default \openshmem
+functions as well as the \openshmem functions to be intercepted, 
+the developers must build the multiple instances of the \openshmem
+functions separately and link them to provide all the definitions. 
+This is necessary so that the developer of the profiling library 
+needs only to define those \openshmem functions that he/she wishes 
+to intercept; references to any other functions will be fulfilled 
+by the default \openshmem library. The link step can be summarized 
+as follows. \\
+
+\noindent\texttt{\% cc ... -lmyprof -lpsma -lsma} \\
+
+Here, \texttt{libmyprof.a} contains the profiler functions that 
+intercept the \openshmem functions to be profiled, 
+\texttt{libpsma.a} contains the name-shifted \openshmem function 
+definitions, and \texttt{libsma.a} contains the default \openshmem
+function definitions. 
+
+\subsubsection{\Cstd[11] Type-Generic Interfaces}
+\label{sec:pshmem_c11_type_generic_interfaces}
+\openshmem currently provides support for type-generic RMA, AMO, 
+and point-to-point synchronization routines through \Cstd[11] 
+generic selection. While these interfaces are defined as macros 
+and mapped to \Cstd interface bindings, they cannot be treated as 
+defined function symbols in the library. Thus, these interfaces 
+cannot have any name-shifted \FUNC{pshmem\_} routines in 
+the profiling library. Furthermore, because of the limitations in 
+\Cstd \FUNC{\_Generic} semantics, the \Cstd[11] fixed-width type 
+bindings may map to \Cstd basic types (e.g. \VAR{int32\_t} may map 
+to \VAR{int}), which also limits the profiling library to 
+have the name-shifted routines with such data types. 
```

Comment:
Suggest: "Furthermore, because no two associations in a \Cstd[11] \FUNC{_Generic} selection expression can contain compatible types, the type name of the \Cstd operation that is invoked may not be identical to the type name of the call's arguments (e.g. \VAR{int32\_t} may map to \VAR{int})."